src/crimson/osd/pg.cc

```diff
@@ -486,48 +486,56 @@ void PG::on_active_actmap()
   if (peering_state.is_active() && peering_state.is_clean()) {
     // loops until snap_trimq is empty or SNAPTRIM_ERROR.
     Ref<PG> pg_ref = this;
-    std::ignore = seastar::do_until(
-      [this] { return snap_trimq.empty()
-                      || peering_state.state_test(PG_STATE_SNAPTRIM_ERROR);
-      },
-      [this] {
-        peering_state.state_set(PG_STATE_SNAPTRIM);
+    std::ignore = interruptor::with_interruption([this] {
+      return interruptor::repeat(
+        [this]() -> interruptible_future<seastar::stop_iteration> {
+          if (snap_trimq.empty()
+              || peering_state.state_test(PG_STATE_SNAPTRIM_ERROR)) {
+            return seastar::make_ready_future<seastar::stop_iteration>(
+              seastar::stop_iteration::yes);
+          }
+          peering_state.state_set(PG_STATE_SNAPTRIM);
+          publish_stats_to_osd();
+          const auto to_trim = snap_trimq.range_start();
+          snap_trimq.erase(to_trim);
```

Comment:
Previously, on interruptions - we would keep trying again (eagain) until success. This would ensure that an erased `to_trim` is actually trimmed. It seems that by removing the eagain case we may **erase** from `snap_trimq` **without** completing the trimming (on interruptions).
We should probably either erase only on success or re-adding the uncompleted `to_trim` on interruption.